### PR TITLE
Asset mixing with fees

### PR DESCRIPTION
**Pull request by findora-crypto**
This PR adds the proof generation part of asset mixing with fees. 

The PR adds two functions, `build_multi_xfr_cs_with_fees` `asset_mixing_with_fees`, that handle the case when an asset type is the fee and a specific amount needs to be charged. 

Sergio (@Serge-Juarez) can add more description of the anonymous fee into this PR thread.

More description about the changes are noted as follows.

### CODE REVIEW

File: poly-iops/src/plonk/turbo_plonk_cs/mod.rs [hunk 1]
```diff
@@ -493,6 +493,18 @@ impl<F: Scalar> TurboPlonkConstraintSystem<F> {
...
+    }
+
+    // Returns a boolean variable that equals 1 if and only if `left_var` != `right_var`
+    pub fn is_not_equal(&mut self, left_var: VarIndex, right_var: VarIndex) -> VarIndex {
+        let (_, is_not_equal) = self.is_equal_or_not_equal(left_var, right_var);
+        is_not_equal
+    }
+
+    // Returns two boolean variables that equals (1, 0) if and only if `left_var` == `right_var` and (0, 1) otherwise
+    pub fn is_equal_or_not_equal(&mut self, left_var: VarIndex, right_var: VarIndex) -> (VarIndex, VarIndex) {
```
Comment (findora-crypto) [inline poly-iops/src/plonk/turbo_plonk_cs/mod.rs:507]: For convenience, we add a function called `is_equal_or_not_equal` that returns about the `is_equal` (aka `diff_is_zero`) and `is_not_equal` (aka `mul_var`). The original `is_equal` function invokes this function as a subroutine. A new function, `is_not_equal`, is added.

File: zei_api/src/anon_xfr/circuits.rs [hunk 2]
```diff
@@ -290,6 +290,113 @@ pub(crate) fn build_multi_xfr_cs(
     (cs, n_constraints)
 }
 
+/// Returns the constraint system (and associated number of constraints) for a multi-inputs/outputs transaction.
+/// This one also takes fee parameters as input.
+#[allow(dead_code)]
+pub(crate) fn build_multi_xfr_cs_with_fees(
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:296]: This is virtually the same as `build_multi_xfr_cs` with only more input parameters and a change below (see the next review comment).

File: zei_api/src/anon_xfr/circuits.rs [hunk 3]
```diff
@@ -290,6 +290,113 @@ pub(crate) fn build_multi_xfr_cs(
...
+    // add asset-mixing constraints
+    let inputs: Vec<(VarIndex, VarIndex)> = payers_secrets
+        .into_iter()
+        .map(|payer| (payer.asset_type, payer.amount))
+        .collect();
+    let outputs: Vec<(VarIndex, VarIndex)> = payees_secrets
+        .into_iter()
+        .map(|payee| (payee.asset_type, payee.amount))
+        .collect();
+    asset_mixing_with_fees(&mut cs, &inputs, &outputs, fee_type, fee_calculating_func);
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:391]: This is the only change from `build_multi_xfr_cs`.

File: zei_api/src/anon_xfr/circuits.rs [hunk 4]
```diff
@@ -642,6 +751,137 @@ fn asset_mixing(
...
+/// 2. Similarly, compute [sum_out_1, ..., sum_out_m] from outputs.
+/// 3. Enumerate pair `(i \in [n], j \in [m])`, check that:
+///         `(type_in_i == fee_type) \lor (type_in_i != type_out_j) \lor (sum_in_i == sum_out_j)`
+///         `(type_in_i != fee_type) \lor (type_in_i != type_out_j) \lor (sum_in_i == sum_out_j + fee)`
+/// 4. Ensure that except the fee type, all the input type has also shown up as an output type.
+/// 5. Ensure that for the fee type, if there is no output fee type, then the input must provide the exact fee.
+///
+/// This function assumes that the inputs and outputs have been correctly bounded.
+#[allow(dead_code)]
+fn asset_mixing_with_fees(
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:776]: We add a new function `asset_mixing_with_fees`. Note that this function has much more complicated logic, which we will now discuss.

File: zei_api/src/anon_xfr/circuits.rs [hunk 5]
```diff
@@ -642,6 +751,137 @@ fn asset_mixing(
...
+                ); // amount
+                cs.add(sum, adder)
+            });
+            (output.0, sum_var)
+        })
+        .collect();
+
+    // Initialize a constant value `fee_type_val`
+    let fee_type_val = cs.new_variable(fee_type);
+    cs.insert_constant_gate(fee_type_val, fee_type);
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:821]: We use a constant gate to ensure that some variables are constants.

File: zei_api/src/anon_xfr/circuits.rs [hunk 6]
```diff
@@ -642,6 +751,137 @@ fn asset_mixing(
...
+        inputs.len() as u32,
+        outputs.len() as u32,
+    ));
+    let fee_var = cs.new_variable(fee);
+    cs.insert_constant_gate(fee_var, fee);
+
+    // At least one input type is `fee_type` by checking `flag_exists_fee_type = 0`
+    // and also check that the amount is matching
+    // and also check that every input type appears in the set of output types (except if the fee has used up)
+    let mut flag_exists_fee_type = cs.one_var();
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs]: `flag_no_fee_type` is the first check. It must be `0` in the end, so that there is at least one input type that is of the fee type.

File: zei_api/src/anon_xfr/circuits.rs [hunk 7]
```diff
@@ -642,6 +751,137 @@ fn asset_mixing(
...
+    for (input_type, input_sum) in inputs_type_sum_amounts {
+        let (is_fee_type, is_not_fee_type) =
+            cs.is_equal_or_not_equal(input_type, fee_type_val);
+        flag_exists_fee_type = cs.mul(flag_exists_fee_type, is_not_fee_type);
+
+        let zero_var = cs.zero_var();
+
+        // If there is at least one output that is of the same type as the input, then `flags_exists_matching_output = 0`
+        // Otherwise, `flags_exists_matching_output = 1`.
+        let mut flag_exists_matching_output = cs.one_var();
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs]: `flag_no_matching_output` is the second check that subsumes a previous but now deleted test that ensures that "for each input type, there is at least a corresponding output type, except in the corner case when the input type is the fee type and it happens that all the fees are spent as transaction fees".

File: zei_api/src/anon_xfr/circuits.rs [hunk 8]
```diff
@@ -642,6 +751,137 @@ fn asset_mixing(
...
+            cs.insert_mul_gate(type_matched_and_is_fee_type, diff_minus_fee, zero_var)
+        }
+
+        // If it is not the fee type, then `flags_exists_matching_output` must be 0
+        cs.insert_mul_gate(is_not_fee_type, flag_exists_matching_output, zero_var);
+
+        // Otherwise, `flags_exists_matching_output * (input_sum - fee_var) = 0`
+        let input_minus_fee = cs.sub(input_sum, fee_var);
+        let condition = cs.mul(is_fee_type, flag_exists_matching_output);
+        cs.insert_mul_gate(condition, input_minus_fee, zero_var)
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:868]: We allow the output to have no fee type, if the amount of the fee type from the input is exactly the fee needed.

File: zei_api/src/anon_xfr/circuits.rs [hunk 9]
```diff
@@ -642,6 +751,136 @@ fn asset_mixing(
...
+        inputs.len() as u32,
+        outputs.len() as u32,
+    ));
+    let fee_var = cs.new_variable(fee);
+    cs.insert_constant_gate(fee_var, fee);
+
+    // At least one input type is `fee_type` by checking `flag_no_fee_type = 0`
+    // and also check that the amount is matching
+    // and also check that every input type appears in the set of output types (except if the fee has used up)
+    let mut flag_no_fee_type = cs.one_var();
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:834]: This is the first check---`flag_no_fee_type` must be zero, which means that there exists an input type that is the fee type.

File: zei_api/src/anon_xfr/circuits.rs [hunk 10]
```diff
@@ -642,6 +751,136 @@ fn asset_mixing(
...
+    for (input_type, input_sum) in inputs_type_sum_amounts {
+        let (is_fee_type, is_not_fee_type) =
+            cs.is_equal_or_not_equal(input_type, fee_type_val);
+        flag_no_fee_type = cs.mul(flag_no_fee_type, is_not_fee_type);
+
+        let zero_var = cs.zero_var();
+
+        // If there is at least one output that is of the same type as the input, then `flag_no_matching_output = 0`
+        // Otherwise, `flag_no_matching_output = 1`.
+        let mut flag_no_matching_output = cs.one_var();
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:844]: This is the second check, which also subsumes a previous check that is now removed. 
- If the asset type is not the fee type, then there must be an output that is of this asset type.
- If the asset type is the fee type, then there can be no corresponding output only if the input for the fee type is exactly the amount needed to pay for the fee.

File: zei_api/src/anon_xfr/circuits.rs [hunk 11]
```diff
@@ -642,6 +751,136 @@ fn asset_mixing(
...
+
+        // Otherwise, `flag_no_matching_output * (input_sum - fee_var) = 0`
+        let input_minus_fee = cs.sub(input_sum, fee_var);
+        let condition = cs.mul(is_fee_type, flag_no_matching_output);
+        cs.insert_mul_gate(condition, input_minus_fee, zero_var)
+    }
+    cs.insert_constant_gate(flag_no_fee_type, BLSScalar::zero());
+
+    // check that every output type appears in the set of input types
+    for &(output_type, _) in outputs {
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:873]: The output check is still necessary because otherwise one can have an output for an asset type that does not exist in the input.

File: zei_api/src/anon_xfr/circuits.rs [hunk 12]
```diff
@@ -1036,80 +1275,693 @@ pub(crate) mod tests {
...
-
-        // compute Pedersen commitment
-        let pc_gens_jubjub = PedersenGens::<JubjubPoint>::new(2);
-        let amount_jj = JubjubScalar::from_u32(71);
-        let at_jj = JubjubScalar::from_u32(52);
-        let blind_pc = JubjubScalar::random(&mut prng);
-        let ped_comm = pc_gens_jubjub
-            .commit(&[amount_jj, at_jj], &blind_pc)
-            .unwrap(); // safe unwrap
+    fn test_asset_mixing_with_fees() {
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:1278]: Similar to `test_asset_mixing`, but with fees and much more comprehensive tests, including the various situations when the fee is enough, too much, or insufficient.

File: zei_api/src/anon_xfr/circuits.rs [hunk 13]
```diff
@@ -1317,6 +2169,64 @@ pub(crate) mod tests {
         test_xfr_cs(inputs, outputs, false);
     }
 
+    #[test]
+    fn test_build_multi_xfr_cs_with_fees() {
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:2173]: This is similar to `test_build_multi_xfr_cs`.

File: zei_api/src/anon_xfr/circuits.rs [hunk 14]
```diff
@@ -1336,4 +2246,27 @@ pub(crate) mod tests {
             assert!(verify.is_err());
         }
     }
+
+    fn test_xfr_cs_with_fees(
```
Comment (findora-crypto) [inline zei_api/src/anon_xfr/circuits.rs:2250]: This is similar to `test_xfr_cs`.

File: zei_api/src/setup.rs [hunk 15]
```diff
@@ -125,13 +125,20 @@ impl UserParams {
...
+        let dummy_fee_func = |_, _| 0u32;
+
         let (cs, n_constraints) = match tree_depth {
-            Some(depth) => {
-                build_multi_xfr_cs(AMultiXfrWitness::fake(n_payers, n_payees, depth))
-            }
-            None => build_multi_xfr_cs(AMultiXfrWitness::fake(
-                n_payers, n_payees, TREE_DEPTH,
-            )),
+            Some(depth) => build_multi_xfr_cs_with_fees(
```
Comment (findora-crypto) [inline zei_api/src/setup.rs:132]: We change the setup code (which would retire in the future after we have our CRS) to set up the parameters sufficient for `build_multi_xfr_cs_with_fees`.

**Review by hridambasu — APPROVED** [review]
Looks Good!

**Review by kingpinXD — APPROVED** [review]

File: zei_api/src/anon_xfr/circuits.rs (hunk 3 above)
Comment (Serge-Juarez) [reply zei_api/src/anon_xfr/circuits.rs:391]: Having a way to indicate which asst type is the target to crarge the fees and a parameter to supply the function which compute such fees is a great point, since it allow findora have a resilient fee implementation.

**Review by Serge-Juarez — APPROVED** [review]
